### PR TITLE
Make GetMemory use MaxBufferSize for MemoryPool

### PR DESCRIPTION
**Pull request by jkotalik**
Another issue with StreamPipeWriter found by https://github.com/aspnet/AspNetCore/pull/7110

### CODE REVIEW

File: src/Http/Http/test/PipeWriterTests.cs [hunk 1]
```diff
@@ -90,13 +90,11 @@ public void CanGetSameMemoryWhenNoAdvance()
...
         [Fact]
-        public void CanGetNewSpanWhenNoAdvanceWhenSizeTooLarge()
+        public void GetSpanWithZeroSizeHintReturnsMaxBufferSizeOfPool()
         {
             var span = Writer.GetSpan(0);
 
-            var secondSpan = Writer.GetSpan(10000);
-
-            Assert.False(span.SequenceEqual(secondSpan));
+            Assert.Equal(4096, span.Length);
```
Comment (halter73) [inline src/Http/Http/test/PipeWriterTests.cs]: Can you verify this same result for when 0 < sizeHint < 4096?

Can you also update CanGetNewMemoryWhenSizeTooLarge to verify that memoryLarge.Length is equal to `MemoryPool<byte>.MaxBufferSize`?

File: src/Http/Http/src/StreamPipeWriter.cs [hunk 2]
```diff
@@ -278,7 +278,8 @@ private void AddSegment(int sizeHint = 0)
             }
 
             // Get a new buffer using the minimum segment size, unless the size hint is larger than a single segment.
-            _currentSegmentOwner = _pool.Rent(Math.Max(_minimumSegmentSize, sizeHint));
+            // Also, the size cannot be larger than the MaxBufferSize of the MemoryPool
+            _currentSegmentOwner = _pool.Rent(Math.Min(Math.Max(_minimumSegmentSize, sizeHint), _pool.MaxBufferSize));
```
Comment (halter73) [inline src/Http/Http/src/StreamPipeWriter.cs]:
```suggestion
            _currentSegmentOwner = _pool.Rent(Math.Clamp(sizeHint, _minimumSegmentSize, _pool.MaxBufferSize));
```

https://docs.microsoft.com/en-us/dotnet/api/system.math.clamp?view=netcore-2.2#System_Math_Clamp_System_Int32_System_Int32_System_Int32_

**Review by halter73 — APPROVED** [review]

File: src/Http/Http/src/StreamPipeWriter.cs (hunk 2 above)
Comment (jkotalik) [reply src/Http/Http/src/StreamPipeWriter.cs]: Yo that API is hype.

**Comment by jkotalik** [comment]
Not sure why the last test run failed:
```
2019-02-01T18:41:16.1474664Z   Total tests: Unknown. Passed: 22139. Failed: 0. Skipped: 52.
2019-02-01T18:41:16.1476790Z   Test Run Aborted.
2019-02-01T18:41:16.1477604Z   Test execution time: 3.6984 Minutes
```

**Comment by davidfowl** [comment]
An abort usually means there's a crash.

**Comment by halter73** [comment]
If it's a debug build, Debug.Assert failures are one of the more common reasons for aborted test runs.

**Comment by jkotalik** [comment]
We didn't get artifacts for test runs. Fix: https://github.com/aspnet/AspNetCore/pull/7251

**Pull request merged**